dart/common/Uri.cpp

```diff
@@ -196,17 +220,129 @@ void Uri::clear()
 }
 
 //==============================================================================
+bool Uri::fromString(const std::string& _input)
+{
+  // This is regex is from Appendix B of RFC 3986.
+  static regex uriRegex(
+    R"END(^(([^:/?#]+):)?(//([^/?#]*))?([^?#]*)(\?([^#]*))?(#(.*))?)END",
+    regex::extended | regex::optimize);
+  static const size_t schemeIndex = 2;
+  static const size_t authorityIndex = 4;
+  static const size_t pathIndex = 5;
+  static const size_t queryIndex = 7;
+  static const size_t fragmentIndex = 9;
+
+  clear();
+
+  smatch matches;
+  if(!regex_match(_input, matches, uriRegex))
+    return false;
+
+  assert(matches.size() > schemeIndex);
+  const ssub_match& schemeMatch = matches[schemeIndex];
+  if(schemeMatch.matched)
+    mScheme = schemeMatch;
+
+  assert(matches.size() > authorityIndex);
+  const ssub_match& authorityMatch = matches[authorityIndex];
+  if(authorityMatch.matched)
+    mAuthority = authorityMatch;
+
+  assert(matches.size() > pathIndex);
+  const ssub_match& pathMatch = matches[pathIndex];
+  if(pathMatch.matched)
+    mPath = pathMatch;
+
+  assert(matches.size() > queryIndex);
+  const ssub_match& queryMatch = matches[queryIndex];
+  if(queryMatch.matched)
+    mQuery = queryMatch;
+
+  assert(matches.size() > fragmentIndex);
+  const ssub_match& fragmentMatch = matches[fragmentIndex];
+  if (fragmentMatch.matched)
+    mFragment = fragmentMatch;
+
+  return true;
+}
+
+//==============================================================================
+bool Uri::fromPath(const std::string& _path)
+{
+  // TODO(JS): We might want to check validity of _path.
+
+  static const std::string fileScheme("file://");
+
+#ifdef _WIN32
+  // Replace backslashes (from Windows paths) with forward slashes.
+  std::string unixPath = _path;
+  std::replace(std::begin(unixPath), std::end(unixPath), '\\', '/');
+
+  return fromString(fileScheme + "/" + _path);
+#else
+  return fromString(fileScheme + _path);
+#endif
+}
+
+//==============================================================================
+bool Uri::fromStringOrPath(const std::string& _input)
+{
+  // Assume that any URI without a scheme is a path.
+  static regex uriSchemeRegex(R"END(^(([^:/?#]+)://))END");
```

Comment:
One possible solution I can think now is:
* For Unix systems, we use corrected regex (```R"END(^(([^:/?#]+):))END")```) so it really checks existence of scheme
* For Windows, we use different regex that checks Windows style path as ```R"END([a-zA-Z]:[/|\\])END"```. We might don't need scheme check here since we assume that incoming path is an absolute path.